File: regression/linear-regression/LinearRegression.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <numeric>

using namespace std;

template <class A, class B>
class LinearRegression {
public:
	A B_0;
	A B_1;
	void fit(vector<vector<B>> dataset) {
		vector<B> X = dataset[0]; //x
		vector<B> y = dataset[1]; //y

		A X_sum;
		A y_sum;
		int n = X.size();

		for (int i = 0; i < X.size(); i++) {
			X_sum += X[i];
			y_sum += y[i];
		}

		A X_mean = X_sum / n;
		A y_mean = y_sum / n;

		A SSxx = 0;
		A SSxy = 0;
		//declaring SSxy
		{
			A t_sum = 0;
			for (int i = 0; i < n; i++) {
				t_sum += X[i] * y[i];
			}
			SSxy = t_sum - n * X_mean * y_mean;
		}

		//declaring SSxx
		{
			A t_sum = 0;
			for (int i = 0; i < n; i++) {
				t_sum += X[i] * X[i];
			}
			SSxx = t_sum - n * X_mean * X_mean;
		}

		B_1 = SSxy / SSxx;
		B_0 = y_mean - B_1 * X_mean;
	};

	B predict(const B& dataset) {
		return B_0 + (B_1 * dataset);
	};

};

File: regression/simple-linear-regression/LinearRegression.py
import numpy as np

class LinearRegression:
    def __init__(self, a, epochs, W=None, b=0):
        self.a = a
        self.epochs = epochs
        self.W = W
        self.b = b

    def fit(self, X, y):
        self.n = np.size(X)
        SS_xy = np.sum(y*X) - self.n*np.mean(X)*np.mean(y)
        SS_xx = np.sum(X*X) - self.n*np.mean(X)*np.mean(X)

        b_1 = SS_xy / SS_xx
        b_0 = np.mean(y) - b_1*np.mean(X)

        return b_0, b_1

    def predict(self, X):
        return self.W * X + self.b
